keps/0028-transient-parameters.md

```diff
@@ -0,0 +1,264 @@
+- Start a backup with a specific name
+- Evict a specific pod from a stateful set
+- Start a repair plan for a specific node or datacenter
+
+All these examples have in common that they require input parameters. The parameters are required, and the user should be forced to set them. If we do not have transient parameters, it might happen that a parameters are still set from a previous invocation and the user forgets to set it. Additionally, using persistent parameters for these use cases would clutter the instance parameter store with values that are not required to be saved.
+
+When parameters are transient, they should be set to their default value after plan execution, and KUDO can error out if a required parameter is not set. 
+
+### Goals
+
+Make it possible to have transient parameters that are only available withing a single plan execution. 
+
+### Non-Goals
+
+- Set parameters to specific values (except for a default)
+- Set or change parameters at other moments than at the end of a plan
+
+## Proposals
+
+#### Persistent vs. Transient parameters
+
+Add an attribute `persistent` to parameter specifications in `params.yaml`:
+
+```yaml
+  - name: BACKUP_NAME
+    description: "The name under which the backup is saved."
+    persistent: "false"
+```
+
+The default value for this flag would be `true`, all parameters that are defined without the attribute are persistent by default.
+
+If the flag is set to `false`, the parameter is transient and its value will not be persisted in the instance resource.
+
+Transient parameters can be easily identified and will only be stored in the `planExecution` section of the instance. As the `planExecution` is reset after the plan finishes, the transient parameters from the plan execution will be discarded as well.
+
+- Transient parameters cannot be `immutable`
+  - They are reset after every plan execution and are therefore mutable by design
+- Transient parameters can have a default value
+- Transient parameters cannot be `required`
+  - This would be a desired property and could be theoretically allowed. But transient parameters are often plan specific, and the `required` scope is global, it would be mostly too wide of a scope.
+
+Pros:
+- Easy extension for parameters from the definition point of view
+- Parameters are not bound to specific plans. Multiple plans can use the same transient parameters. (For example `backup` and `restore` could use the same `BACKUP_NAME` parameter)
+
+Cons:
+- The parameters for a specific plan are not separated from persistent parameters - It can be easy to miss the attribute.
+- Without plan specific parameters, it is hard to determine if a plan requires a specific parameter. It will be easy for a user to forget to specify a transient parameter, and `kudoctl` will not be able to determine that a parameter is required to render a plan-specific resource.
```

Comment:
Tbh, at the moment we don't have a _parameter-to-plan_ mapping either. We may have a `trigger` field on the parameter, but the param can be used in any other plans as well, so I am not sure how _strong_ our mapping here is anyway.

The more I think about it, i'd rather get rid of the trigger and always specify plans explicitly. But that's a discussion for a different KEP.

Reply:
The `trigger` is always there, it just [defaults to deploy](https://github.com/kudobuilder/kudo/blob/main/pkg/webhook/instance_admission.go#L349) if not specified. And to be precise:
- one parameter can be used in many **templates** so 1:n 
- one parameter can trigger only one plan so 1:1
- one plan can be triggered by many parameters so 1:n

Reply:
Yeah - the "default to `deploy`" is problematic as well.

There is a relation missing:
 - one *template* can be used in many plans so 1:n


I feel this relationship is flawed. If one parameter can be used in a template, and that template can be used in multiple plans, why does the parameter trigger a single plan?


One of the next things I'll KEP out is probably the "Define No-Plan Trigger" on parameters, because with the `BACKUP_NAME` parameter we get a problem:

`BACKUP_NAME` is used by `backup` plan and `restore` plan. If the `BACKUP_NAME` parameter has the `backup` plan as the trigger, it can't be used together with the `restore` plan - because specifying multiple parameters that trigger different plans is prevented by the webhook...

Reply:
Or maybe it's enough to say "If a plan is explicitly triggered, it doesn't matter what `trigger` attributes the parameters have" - that might work as well...